python_modules/libraries/dagster-components/dagster_components/core/defs_module.py

```diff
@@ -0,0 +1,279 @@
+    component_file_path = path / "component.yaml"
+    source_tree = parse_yaml_with_source_positions(
+        component_file_path.read_text(), str(component_file_path)
+    )
+    return (
+        source_tree.source_position_tree,
+        _parse_and_populate_model_with_annotated_errors(
+            cls=ComponentFileModel, obj_parse_root=source_tree, obj_key_path_prefix=[]
+        ),
+    )
+
+
+@record
+class DefsModuleDecl(ABC):
+    path: Path
+
+    def get_source_position_tree(self) -> Optional[SourcePositionTree]:
+        return None
+
+    @staticmethod
+    def from_path(path: Path) -> Optional["DefsModuleDecl"]:
+        # this defines the priority of the decl types, we return the first one that matches
+        decltypes = (
+            YamlComponentDecl,
+            PythonComponentDecl,
+            PythonModuleDecl,
+            SubpackageDefsModuleDecl,
+        )
+        decl_filter = filter(None, (cls.from_path(path) for cls in decltypes))
+        return next(decl_filter, None)
+
+    @staticmethod
+    def from_module(module: ModuleType) -> Optional["DefsModuleDecl"]:
+        """Given a Python module, returns a corresponding defs declaration.
+
+        Args:
+            module (ModuleType): The Python module to convert to a component declaration node.
+
+        Returns:
+            Optional[ComponentDeclNode]: The corresponding component declaration node, or None if the module does not contain a component.
+        """
+        module_path = (
+            Path(module.__file__).parent
+            if module.__file__
+            else Path(module.__path__[0])
+            if module.__path__
+            else None
+        )
+        return DefsModuleDecl.from_path(
+            check.not_none(module_path, f"Module {module.__name__} has no filepath")
+        )
+
+    @abstractmethod
+    def load(self, context: ComponentLoadContext) -> DefsModule: ...
+
+
+@record
+class SubpackageDefsModuleDecl(DefsModuleDecl):
+    """A folder containing multiple submodules."""
+
+    subdecls: Sequence[DefsModuleDecl]
+
+    @staticmethod
+    def from_path(path: Path) -> Optional["SubpackageDefsModuleDecl"]:
+        if path.is_dir():
+            subdecls = (DefsModuleDecl.from_path(subpath) for subpath in path.iterdir())
+            return SubpackageDefsModuleDecl(
+                path=path,
+                subdecls=list(filter(None, subdecls)),
+            )
+        else:
+            return None
+
+    def load(self, context: ComponentLoadContext) -> SubpackageDefsModule:
+        return SubpackageDefsModule(
+            path=self.path,
+            submodules=[decl.load(context.for_decl(decl)) for decl in self.subdecls],
+        )
+
+
+@record
+class PythonModuleDecl(DefsModuleDecl):
+    """A python module containing a `definitions.py` file."""
+
+    path: Path
+
+    @staticmethod
+    def from_path(path: Path) -> Optional["PythonModuleDecl"]:
```

Comment:
I am a little iffy on the `from_path` method just silently returning `None` vs e.g. raising, it feels like an actual exception should occur if you explicitly direct `PythonModuleDecl` to construct itself from an empty folder. This is mostly a stylistic qualm since I think type checking prevents anything bad from actually happening here